Add Events API/Support for tt_metal for host->CQ and CQ<->CQ synchronization #5529

kmabeeTT opened the pull request:
Changes are functional and cleaned up, got two things to follow up on

- [x] Still WIP a FIXME in code for known bug for 2 hw cq, and need to test it (will update this PR code tomorrow)
-  multi-device won't work in current implementation, no API exists for chip-to-chip reads from kernel, need to revisit soon (after this merge).

Tag @tooniz for initial review in the meantime if you have time.

kmabeeTT committed:
#5529: Implement EnqueueQueueRecordEvent(cq, Event) in tt_metal

 - Lightweight cmd that goes through device and updates completion queue,
   returns updated Event struct to caller
 - Intended to be used with EnqueueQueueWaitForEvent(cq,Event) or
   EventSynchronize(Event) for device or host to wait on completion
   of an Event, respectively.
 - Move Event struct population to separate func to be called earlier
   to work with async queues.

kmabeeTT committed:
#5529: Implement EventSynchronize(Event) for Host->CQ sync tt_metal

 - Allows host to block until an Event is completed on a HW CQ

kmabeeTT committed:
#5529: Implement EnqueueQueueWaitForEvent(cq, Event) for CQ->CQ sync in tt_metal

 - Allows a HW CQ to block until an Event is completed on another HW CQ
 - Add assert that multichip/cross-chip sync not supported currently
 - More multi-cq Events tests

tooniz commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1144,6 +1244,55 @@ void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<
 }
 
+void EnqueueQueueRecordEvent(CommandQueue& cq, Event &event) {
+    TT_ASSERT(cq.id() == 0, "EnqueueQueueRecordEvent only supported on first command queue on device for time being.");
```
Is this restriction necessary? Why can't we support recording events on CQ1?

kmabeeTT replied:
Not necessary, but keeping until I can test with multi-CQ (WIP today, hope to include with this PR).

tooniz commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -1144,6 +1244,55 @@ void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<
+
+
+void EnqueueQueueWaitForEvent(CommandQueue& cq, Event &event) {
+    TT_ASSERT(cq.id() == 0, "EnqueueQueueRecordEvent only supported on first command queue on device for time being.");
```
Same, is this restriction necessary? Why can't we support recording events on CQ1?

kmabeeTT replied:
Not necessary, but keeping until I can test with multi-CQ (WIP today, hope to include with this PR).

tooniz reviewed (approved):
Feedback regarding CQ1 not blocking this merge, but it'll need to be addressed before doing multi-CQ testing with events

kmabeeTT commented:
Hi @TT-billteng , @pgkeller , @abhullar-tt this is all tidied up, can I get your approval on these changes for #5529 please?  This is initial implementation of 3 Events API's to be used for inserting and syncing on Events from Host or Device (CQ).

- implementation, tested with synchronous hw CQ's (num_hw_cq=1, 2) for single device
- new gtests (still WIP on another 1 or 2 to include before this gets merged)
- remove redundant tt_eager APIs with same scope/name as tt_metal after chat with Akhmed.

Restriction (going to add this comment in parent ticket):

- Won't work with multi-device today, added an assert to catch if user uses `EnqueueQueueRecordEvent(cq0..)` and `EnqueueQueueWaitForEvent(cq1...)` where cq0 and cq1 are on two different devices.  Needs to be discussed next, and revisited after multi-device FD makes bit more progress. May need to change to have "last_completed_event" stored in hugepage per CQ (but strongly prefer not to, due to perf impact).

Some thoughts/decisions along the way in case it helps with review:

- matches interface discussed in parent ticket #5529
- I purposely kept this implementation simple, light. Thought about having a program/kernel to do the "on device wait" but realized it's just a couple lines of code to have the while-loop in the dispatcher kernel.
- To facilitate "Event Sync" I just pass 4 args to dispatcher kernel in 8 bytes (event_id, noc_x, noc_y, sync_ena) to tell dispatcher that this is a sync cmd and it should poll some noc_xy core at known address (reserved for "last completed event ID") until desired event_id (or greater) is seen.
- Reserved an extra slot in common runtime adress map in the gap that is used for dispatch core to store "last completed event_id" for that sync to work.
- Similarly, on host, added an extra state variable per CQ to store "last completed event id" for the host side sync.

pgkeller commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:
```diff
@@ -0,0 +1,57 @@
+
+    // Sync on some earlier events per CQ just to show it works.
+    for (uint i = 0; i < cqs.size(); i++) {
+        EventSynchronize(sync_events.at(i)[3]);
```
3, 4 made me pause, both are <10, maybe use constexpr for them and assert they are less than num_events?

kmabeeTT replied:
Yeah, that's fair. Not worth the trouble to come up with some funky expression, I just changed it to loop over all events per CQ again, and sync on them in order per CQ (instead of bouncing between CQs), pushing in few mins after rebase/build tested locally.

pgkeller commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:
```diff
@@ -53,5 +53,81 @@ TEST_F(MultiCommandQueueSingleDeviceFixture, TestEventsEventSynchronizeSanity) {
+    }
+}
+
+// TODO/WIP - More tests, mixing reads/writes potentially and using events to synchronize Read-After-Write.
```
tests above appear to cover well that events happen in order alone and across queues, but not clear we have a test that ensures that events wait until, eg, a kernel completes.  wonder if you could run something that runs long and returns some data and then afterwards do a host write to trash that location of L1.  if the event completes early, then the data returened be corrupt.  or maybe this is covered somehow?

kmabeeTT replied:
Yep, I have other tests below and newly added after this and WIP that will further test multi-cq ordering, this one was purposely kept ~simple.

pgkeller commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp:
```diff
@@ -128,6 +129,94 @@ TEST_F(MultiCommandQueueSingleDeviceFixture, TestEventsEnqueueQueueWaitForEventC
+            bool local_pass = (srcs[i] == result);
+            pass &= local_pass;
+        }
+    }
```
ok, this test covers ordering, cool

pgkeller reviewed (approved):
recommend review by Almeet and/or Andrew

abhullar-tt commented on tt_metal/host_api.hpp:
```diff
@@ -397,15 +398,35 @@ void EnqueueTrace(Trace& trace, bool blocking);
+ * |--------------|------------------------------------------------------------------------|-------------------------------|------------------------------------|----------|
+ * | event        | The event object that host will wait on for completion.                | Event &                       |                                    | Yes      |
+ */
+void EventSynchronize(Event &event);
```
For docs to pick up these APIs you will need to add them under `docs/source/tt_metal/apis/host_apis/command_queue/`

kmabeeTT replied:
Whoops yes thanks good call - made updates and tested them out locally with web server.

abhullar-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -904,6 +969,39 @@ void HWCommandQueue::enqueue_program(
+    event_ref.event_id = this->manager.get_next_event(this->id);
+    event_ref.device = this->device;
+
+    if ((this->manager.get_issue_queue_write_ptr(this->id)) + command_size >= this->manager.get_issue_queue_size(this->id)) {
```
I think this should be checking that issue_queue_write_ptr + command_size >= issue queue limit

kmabeeTT replied:
Whoops yes I copied buggy version from enqueue-program path, thx for flagging, fixed, about to push.

abhullar-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -904,6 +969,39 @@ void HWCommandQueue::enqueue_program(
+    ZoneScopedN("HWCommandQueue_enqueue_wait_for_event");
+
+    uint32_t command_size = DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND;
+    if ((this->manager.get_issue_queue_write_ptr(this->id)) + command_size >= this->manager.get_issue_queue_size(this->id)) {
```
same as above, should be checking against issue queue limit

abhullar-tt commented on tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:
```diff
@@ -288,6 +298,20 @@ FORCE_INLINE void wait_for_program_completion(uint32_t num_workers) {
+FORCE_INLINE void wait_for_event(uint32_t event_id, uint32_t noc_x, uint32_t noc_y) {
+    uint64_t src_noc_addr = get_noc_addr(noc_x, noc_y, CQ_COMPLETION_LAST_EVENT);
+    do {
+        noc_async_read(src_noc_addr, CQ_COMPLETION_16B_SCRATCH, 4);
```
the `_SCRATCH` suffix is a little misleading because it isn't really being used as scratch space, it seems to be a sync location so maybe this can be renamed to `CQ_COMPLETION_SYNC`?

kmabeeTT replied:
We talked offline, I'm treating it as a scratch region that could be used as destination for any type of 16B noc read to L1 by this dispatcher core, but just using it here for the first time to read back the last-completed-event-id.

kmabeeTT commented:
Passing CI before merge at TOTT:

https://github.com/tenstorrent-metal/tt-metal/actions/runs/8089432962
